### PR TITLE
Create dedicated shader functions for PBR lighting

### PR DESCRIPTION
**Pull request by ptrgags**
This PR adds new shader builtins to handle PBR lighting. This allows direct PBR calculations outside of `processPbrMaterials.js`. In a future PR, a refactored `Model` will use these functions as part of a larger shader.

The following builtins were added:

* `czm_pbrParameters` -- struct for the common material parameters for both PBR workflows
* `czm_pbrMetallicRoughnessMaterial()` -- function for preprocessing the metallic/roughness workflow settings. This assumes the textures are already read and converted to linear space
* `czm_pbrSpecularGlossinessMaterial()` -- similar, but built for the specular/glossy workflow
* `czm_defaultPbrMaterial()` -- default settings for the `czm_pbrParameters` struct.
* `czm_pbrLighting()` -- this handles the actual lighting calculations.

Local Sandcastles:
* [Metallic Roughness Panels](http://localhost:8080/Apps/Sandcastle/index.html#c=hVVtb9MwEP4rVvmSouImexGs7SZgK9OkVZu2MvGhErjJpbNw7Mh2uhW0/845ifOygvhQJ2c/zz1357s0VtJYsuXwBJqcEglP5BwMLzL6UO4Fq0Fc2udKWsYl6NVgOF3JlYxLpnlkScn8gTu/su8Zs6A5E8QZG7CL2g66h1cyLyzpWUPyeyUJeSNQghyGYeisnsMMRWqnF5CyQrS++56mjup+W4gPSLFFXg9AjS0TQLUEUqd3e0UOaXQUHZ8cVAepUAwDBMuE4DE6wA2lgyMakrfoke6GZEzQmrZgrYrNowRjGnQUevizgzuz1q1Kh+Edksub6wtkuPcgomE4IiF9H7n14KTKpAP+fP117sEIcKhDt9TIErNmBs6VUO5OMv4cOM6olBmRXD0FPinHi4ZDHxEWNl/rW6YZAkAbkrevp/54UXPvfK7NDTgfpNUeVXajVZlNiZzplcug5d4dSaUzJubnbWKizRyrFNVLJ3PBN4/Wp+6RZSn92gHniHnXl8yV4ZYruVTzHXSVMfO2BtdOhctNnXJe5ybrp/DPJph6o62nz550W3C5KAUYDttSSchYnqOIQ2Q04WlaGMAYlosAQ6jS0GALLUmGxstK/tibSX83/alubqwct5StNY8n1ewRYlShY5jUDtween7pjruG2DK5ES6Y6qNBTQwSaK55htXbgqEsScridFRv/WlQK21AYTH07gqdMolJT7rwy1enntXyevA7H5TntXjSRjzZg9NUq+wCNhrABC2DkHfRB9dk3a3/2UfHNGztYedsC9rC8xfXz7aJYY5jkRvFk/tCp3jpn/IcsEEwV/owv1vOv33/cnO3+LRs/Lx4l80Layi9Yvzbc1sVtlZbuMRxlMkEe0AY6Dt/Ga5kdetN9WgrR7O2s/qtNh2MBjNjdwLOvNRHnuVKW1JoEVA6tpDlAtFmvC7in2BpbEzVzoTMxl3qLOFbwpPTv/z/kFgwY/AkLYS4579gNTibjRG/R8Uvc4JjdIN3INjOwR6js+tqk1I6G6P5d6ZVSqyZfuX5Dw) image
* [Metallic Roughness with Textures](http://localhost:8080/Apps/Sandcastle/index.html#c=jVZdb9s2FP0rhPdQaXApf6RoYyfBusQNAsRIkLrFHgy0tHRlC6NIgaScuIP/+y4lkZLiDutDJF3ynHM/SSeWQhuyz+AZFLkkAp7JNeiszOnXai1YD+LKvpbCsEyAWg/C+VqsRVwx9Y4lFfM7rvzIv+XMgMoYJ9bYglk2dtDdvBNFaUjPCsk/a0HIbxxdkOloNLJWTzBHJ43oDaSs5K12X2luqfZvD/GElHvk9QBUmyoB9JZAav093pEpHZ+N351P6o0oSrlkGCIYxnkWowQuSBWc0RH5HTXpISQRQWvehStZbncCtPb48cgRXizBmnPnoi4gBjkltw/3N8ix38GYjkZDMqLvx/Y5OQ/nJ/A/778sHBwhFje1D4+tUBum4VpyabuTZy+BZQ0rV0NSyOfAJWeZ47CmnhANvJhSweQm8KuremmIWYVUbTcVscux4t4e1mFO6CQMm9RPats6cYt9H/OW1S1xS/OrJzw3RsVGPTLFUB+UJkX7eem2l43rJ6flx8tqdDIc1rYvX236EKzpMq3KKU4GUEiVM764bkvO23Zi88fNo9MTnm13xhW4QU5946cdZIGAt31/hdSZyaRYycUBum4x7bYA99ZFJrZNvkWTmGje3L19JM1CW0yXOukertWycsDwGllJATkrCnRiETlNsjQtNWAMq2WAIdRpKMAeCpKjcVyL7ye3jWtM/77y7aoukpRtVBbP6luFEC1LFcOsEWgCL0WWYie0R3Wa3AzSjLzZGVPMoojLmPGd1GZ2dv7+PCrwLjSQVEMTeRYtxPbN0Im9GuZf03KkvtTrAf81Lc+qxGqto33h49i9xRXEhoktt52ofwuojkEALVSW4+jsQVOWJNVkdEr+6HaDpoBbkBi/OtyhKBPY8VkXfvtqN/Bld7we/MkF5XgtnrQRz07gNFUyv4GtAtBByyDk7fiDPV7dpf+zz97RUWuHnb09KLx/PtmTbHwMC+xcoWWWfC5VihP/sSgATwfmSr8unlaLv759enhaflx5naOT9B/MU3rF+G/ltipsI/dwi00XyQwPANfQFz+Ga1F33VePtu5o3h6r/jmbD4aDC20OHK6cqz+yvJDKkFLxgNLIQF5wROtoU8Z/g6Gx1vVZJuQi6lIvkmxPsuTyJ/9WkJgzrXEnLTn/nP2A9eDqIkL8CRV/CRK8Qx6wB5wdLGw3vrqvFymlFxGaP2caKfmGqVfK/wI) -- note that this requires the `paintedMetal` textures I made (see zip below) 
image
* [Specular Glossy Panels](http://localhost:8080/Apps/Sandcastle/index.html#c=hVVbT9swFP4rVveSbsVNuUijFDRWOlSpFYh2aA+VwE1OijXHjmynUCb++45zD2XaQ92c4+/7zs1OAiWNJVsOz6DJOZHwTMZgeBrT+8znrTpBZo+VtIxL0KtO92wlVzLImOaJhRnzET2v8UPMLGjOBHHGBuy8sL3m5lQmqSUtq0v+rCQhnwSGIEe+7zurJRhjkEL0CiKWilq7rXTmqO63heCQpFvktQDU2KwAjBZC5OLdTskRHRwPTk4P841IKGbJRihjcN8YlECX0t7Apz75jKL0pUv6xJmFVt4ODHlEbmeXi+V0jCRneog57u6jFtPZ/eSuBA2o7/dIveZVNODfZz8nteJhj/j0yC2nOTLDhDyKUgNjJZQbCSLIlyqHnst65yh+g2ISCFLBdMmJ+YtX5N8rUuw1GlHVgYNI1vqWaRYD9taQpH48L7cXhfh1xa9G5kRIK+Fe7molVPjq+M4uc8jyl3vTlUrHTEzGdY2iMYmsu9nSaILgmydbdqBEDrJWFWsDnCDmoB0yUYZbruRSTXbQjIw9qLsxc1G43BS1J0VxsvgX5X+VTOGoO1tWT5qHdznPAjC8pkslIWZJgkEcIqZFfzGH5dzDFPIyNNhUSxKj8baSj3u3uRxS+31QjS67qBFbax4M81uLU1OpDmBYCDgfKr81XxQaAsvkRrhk8tcNNQFIoInmMXZvC4ayMMya04h6W+56RaQNKGyG3k1RlEksetiEX7/bLVk1rwW/K5MqeTWe1BkP9+A00iq+go0GMF7NIORg8NUdsqbrf/bxCfVru9vY24K28PLDnWdb5TARgidG8XCR6giHfpkkgAcEa6V4V5eTXw8/bu7ml8tK562UrB5YRWk149/KdVfYWm3hWqtUhkM8A8JAW/ytu5L51Kvu0TocjeuT1T5qZ51eZ2TsTsBFGeobjxOlLUm18CjtW4gTgWjTX6fBb7A0yN9GDjrqN6mjkG8JD88/+HKRQDBjcCdKhVjwV1h1LkZ9xO9R8QMQ4jW6wRkItnOwp8HFLHdSSkd9ND9mWqXEmul3yn8B) 
image
* [Specular Glossy with Textures](http://localhost:8080/Apps/Sandcastle/index.html#c=jVVtT9swEP4rVvlAKhW3wKaNUtAQFIQEAkGF9qESuMkltebYke2Ulan/fZfEiRPKND40yZ3vee7dDZU0lqw4vIImJ0TCKzkHw/OUPpW6YN4LS/lcScu4BD3v9Y/nci7DEmmWLCqRL6h5S59TZkFzJkghJGBvnRy0D69lllvSkfrkz1wSsiPQBTkcjUaF1CFM0YkjvYCY5cJzd5mOC2jxW0F4QPIV4joG1NgygdLgkEQ8jnMDaGXht801HFwETjerFAMk6VOdLI4bkMkgzAXTHVSt/BAWC8UsSYQyBlM0poP06i7WxVmknS30PdMsBczEkMx/ntTHj879VUPWFKggIXWmg0qqo3WiD6GQ++0Kya0KSqVTJqbnviACbYqPYERHA7LvHv2WAU+W9lwJpb3lwYDUj5ZlhgZ7XX+ZMtxyJWdquoa2W0zbF+CmcMFl4tLNXGbSvUX9biJxCl/MOvVyfHYiiItpnN2WDhjuwUxJSFmWoZPCIqV+eGa3AYZQpaEBeyhJisJmLl+21qXuS3fhmm6VmxCzhebhuFoL7JbKdQhjR+ACzyWPsROmsWp67MZoTHaX1mbj4VCokImlMnb85ejb0dC0Z2U9dCiayWR3UFO9G+fPctVil2xrwj9L54ElYcW3KV742LSvIg2hZTIRRTeqC42aECRmpXmK47MCQ1kUldPRKvt9fRq4IiagcBr0+hpJmcSuj9vmV+9Og6b0Na5j/lAHVeO8PfERj7fMaaxVegGJBjCBRxCyt/+9WLG26n/yl6905OV+62wFGi+hy2KbbRPDVAieGcWjx1zHOPVnWQa4IZgrfZo+zKY/ny/vHm7PZg3PpqZsPlgD6RTj38y+KmyhVnClVS6jMS6BaO4rR77pz2XV9aZ61LujqV+t7q4d9wa9ibFrAae1qx88zZS2JNcioHRoIc0EWpvhIg9/gaWhMdU+EzIZtqGTiK8Ij04++G8koWDG4EmcC/HI32DeO53gfq22oPh/EOE9coc9EGxdmC33T28qJaV0MkTxY6RVSiyYfsf8Fw) -- requires the `specularGlossy` textures (see zip below). image

For the above sandcastles I tried creating my own PBR textures by hand so I learn more about how this works. It would be good to try this with other PBR textures as well

[pbr-textures.zip](https://github.com/CesiumGS/cesium/files/6474752/pbr-textures.zip)

I've been hosting these images locally with a static server

@lilleyse could you review?

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @ptrgags!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/pbr-lighting/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/CesiumGS/cesium/tree/master/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

**Comment by ptrgags** [comment]
Silly me, in my sandcastles, I was forgetting one super important line:

```
viewer.scene.highDynamicRange = true;
```

[Example](http://localhost:8080/Apps/Sandcastle/index.html#c=hVVdb9MwFP0rVnlJoLjNNgS03QRsZZq0alMpiIdK4CY3rYU/ItvpVtD+O9f5DgXxsDTXPvece659s1gr68iewwMYck4UPJBLsDyX9EuxFqwHcRFfauUYV2DWg3C6VmtV5lAbgwK649vd1UExyeMlU1tAKmdyKIBxIWF3LCkkvuPKT/lNMgeGM0F8sAW3qOKgu3mjstyRXhSSX2tFyDOBtZDT8Xjsox6hRJGK9ApSlouWu8809an+bw/xCcn3mNcDUOsKA6iWQOr17m/IKY3OoldvT8qNVGiGBYJjQvAYCXBBm+CMjslzZKSHkIwIRtMWbHS+3SmwtkFH4xr+6OE+LPBdpGSPQRMPCUKisCqu7C96OCXXd7dXCPbvQUTHYw98HfnnydvSbgf84fbzvAYjwKNO/aNCFpgNs3CphfYHJ/lj4HOGhcyQZPohqJ37vChsKsLuZxtzzwxDABhLsvb1vN5eVLnL2lZzTJ6DtNrDMm60yrDphg9r5aJodXSQShvJxPyyNSZa59ilqHp0nAu80a62XiOjsvNt/ytwhpiXfclMW+64Vis9P0BXGZ23Pbj1KlxtK8tZ5U1Vv6L+bYqpFtp+Hrnnm663cDoaeS0u2RY+YEeTRrKov3u7V4uiLIYDv9IKJMsyxHmEpAlP09z6uV4tAm/hhRcqW2DA5UYRicHTWn0/Gvr6XPvfl+a0i3lO2cbweFIONyFW5yaGSUXg15D5KexQG4gdfmmEL6n3KcoMl9j5PVjKkqRobEf1vt4NKqUtaGykOdwgKVNofdKFX/+xW2e1eT34si6qzmvxpK14cgSnqdHyCrYGwAZtBiEvozf+gnaX/hefvaLjNg47e3swDh4/+llwTQ1zHKnMap58yk2KR/8+ywAvF3qlX+bL1fzrt493y8X7VcPzVFM2L6xJ6TXj38xtV9hG7+EaR1klE7wDwkKf/Clcq/LUm+7RVo7K9mb1r9p0MBzMrDsIuKil3nGZaeNIbkRA6ciBzASi7WiTxz/A0dja8joTMht1U2cJ3xOenP/lPyGJBbMWd9JciE/8J6wHF7MR4o9S8dOf4DDd4RkIdvCwXXRxWy5SSmcjDP+e6bQWG2b+YP4N)
image -- much brighter

Also note that ambient lighting is not handled by this PR, that will be done separately in the Model shader in a future PR.

**Comment by midnight-dev** [comment]
Looks like a nice enhancement for the renderer! I'll have to revisit this when my team wanders back into Cesium territory. 

I'm curious about the naming convention. Is the `czm_` prefix there to prevent collisions with known or likely shader elements? Or does it signify that this PBR technique is substantially different than otherwise standard PBR?

For now, I'm assuming it's just to keep convention with many Cesium labels that have existed since legacy days to help prevent scope pollution. But I don't regularly browse glsl files. Just want to make sure I'm not missing something with the `czm_` prefix 👌

**Comment by ptrgags** [comment]
@midnight-dev PBR shading is already available for glTF models (see

To answer your question about the `czm_` prefix, that's just our standard prefix for naming built-in functions. See the [Coding Guide GLSL Naming section](https://github.com/CesiumGS/cesium/blob/master/Documentation/Contributors/CodingGuide/README.md#naming-1). This is pretty standard PBR stuff.

This PR is just a refactoring of https://github.com/CesiumGS/cesium/blob/master/Source/Scene/processPbrMaterials.js which handles [glTF PBR materials](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#materials) which are based on the metallic/roughness workflow. This file also handles the `KHR_materials_pbrSpecularGlossiness` extension. We're just moving it to a built-in function so it's more reusable.

**Comment by midnight-dev** [comment]
> This PR is just a refactoring ... 

And so it is... I somehow missed PBR as a feature even though there's an existing Sandcastle example. 

Thanks for the information. I will see myself out. ・_・

File: Source/Shaders/Builtin/Functions/pbrLighting.glsl [hunk 1]
```diff
@@ -0,0 +1,95 @@
+const float M_PI = 3.141592653589793;
+
+vec3 lambertianDiffuse(vec3 diffuseColor) {
+  return diffuseColor / M_PI;
+}
+
+vec3 fresnelSchlick2(vec3 f0, vec3 f90, float VdotH) {
+    return f0 + (f90 - f0) * pow(clamp(1.0 - VdotH, 0.0, 1.0), 5.0);
```
Comment (lilleyse) [inline Source/Shaders/Builtin/Functions/pbrLighting.glsl:8]: Out of curiosity are there any notable differences between the shaders here and the shaders in `processPbrMaterials.js` or are they basically the same?

Comment (ptrgags) [reply Source/Shaders/Builtin/Functions/pbrLighting.glsl:8]: The shaders here are essentially the same as the PBR portions from `processPbrMaterials.js`. I did cross reference with a few other sources to check for differences, but the calculations are as described in the glTF spec, the sample viewer, and [this tutorial](https://learnopengl.com/PBR/Lighting) which was helpful in understanding the math

**Comment by ptrgags** [comment]
@lilleyse updated!

File: Source/Shaders/Builtin/Functions/pbrLighting.glsl (hunk 1 above)
Comment (lilleyse) [reply Source/Shaders/Builtin/Functions/pbrLighting.glsl:8]: Great!

**Pull request merged**